### ISSUE TITLE
Different Settings types?

### ISSUE THREAD

**Issue by phadej** [shayne-fletcher/ghc-lib-parser-ex#106]
Using GHC-9.0,1, ghc-lib-parser-9.0.2.20211226 and ghc-lib-parser-ex-9.0.0.6

Is there are a guide how to not run into problems like these?

```
    • Couldn't match expected type ‘GHC.Settings.Settings’
                  with actual type ‘ghc-9.0.1:GHC.Settings.Settings’
      NB: ‘GHC.Settings.Settings’
            is defined in ‘GHC.Settings’
                in package ‘ghc-lib-parser-9.0.2.20211226’
          ‘ghc-9.0.1:GHC.Settings.Settings’
            is defined in ‘GHC.Settings’ in package ‘ghc-9.0.1’
```

and the code is

```haskell
import GHC.Driver.Session   (DynFlags, defaultDynFlags, xopt_set)
import Language.Haskell.GhclibParserEx.GHC.Settings.Config (fakeLlvmConfig, fakeSettings)

fakeDynFlags :: DynFlags
fakeDynFlags = defaultDynFlags fakeSettings fakeLlvmConfig
```

Is there are a guide how to not run into problems like these?
I find it surprising that `ghc-lib-parser-ex` depends on `ghc` lib, it's *ghc-lib-parser*-ex! Do I need to replicate flag logic to depend on `ghc` lib myself too? duh.

**Comment by shayne-fletcher** [comment]
ghc-lib-parser-ex tries to make a "smart" (default) decision as to whether it should link ghc-lib-parser or ghc native libs (basically it's optimized for Hlint).

The behavior is controlled by the flag `auto` with default value `true`. This [blog post](https://blog.shaynefletcher.org/2021/02/configuring-cabal-build-flags.html) details how to force the link according to your needs.

IIUC, I believe setting `auto` to `false` together with `no-ghc-lib` also set `false` will ensure ghc-lib-parser is linked by ghc-lib-parser-ex.

I agree that docs are lacking on this point. I'm sorry you were inconvenienced.

**Comment by phadej** [comment]
I'd preferred either
-  "just use `ghc` when versions match" in `ghc-lib-parser`
- or alternatively some extra shim/compat package,

Then `hlint` could depended `ghc-lib-parser` (or `ghc-shim`) and `ghc-lib-parser-ex` instead **of duplicating the logic itself too** in https://github.com/ndmitchell/hlint/blob/master/hlint.cabal#L83-L90

So I guess I have to duplicate the logic myself, or vendor parts of `ghc-lib-parser-ex`.

TL;DR, if you find copying conditionals in `.cabal` files, maybe you should create a compatibility package instead.
